Revise tree pattern used to match inline finalizable test

hzongaro opened the pull request:
The pattern used to match an inline finalizable test might be incorrect in 64-bit mode as tree simplification might result in 32-bit operations being used.  This change accounts for that possibility by checking whether an `l2i` was applied to the result of the load of `isClassAndDepthFlags`, and checks for 32-bit operations accordingly.

Fixes:  Issue #4885

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

fjeremic commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -3742,32 +3742,60 @@ static bool isFinalizableInlineTest(TR::Compilation *comp, TR::Node *candidate,
+   TR::Node *loadNode = castDownToInt ? r11->getFirstChild() : r11;
 
    /*
       Looking for a pattern of:
```
FYI if you did not know we have quite a handy utility `TR_PatternBuilder` which is a very concise and readable way of doing tree pattern matching. There are several examples if you search the code base (eg. [1]).

[1] https://github.com/eclipse/omr/blob/88fd3ecdbbdc0ef7ace859dacc069b2c38e9ef27/compiler/codegen/OMRTreeEvaluator.cpp#L363-L380

hzongaro replied:
Thanks, Filip!  I did not know, and I had been thinking that something like that was needed.  Why doesn't its use seem to be particularly widespread?

fjeremic replied:
Probably because not many people know about it. I've been encouraging it's use in PR reviews if I spot an opportunity. If it works well for you perhaps you can advocate for it as well? Thanks!

hzongaro replied:
@fjeremic Filip, I was playing around with the `TR_PatternBuilder`.  I think in this case using it might be a bit messy because we want to match either `ificmpne` or `iflcmpne`, `iand` or `land`, etc.  So either I would have to build the patterns with `pb.choice` throughout, or build three different patterns:  one pure `int` case, one pure `long` case and one mixed with an `i2l` in the middle.  We might want to think about having some built-in mechanisms for matching either `int` or `long` operations to simplify this kind of scenario.

The other thing I found is that tree pattern matching seems to be currently disabled.  We'll investigate why that is, but Andrew @andrewcraik  suggested I leave it aside.  If you don't have any objection, that's what I'll do for now.  I will definitely look for further opportunities to use the pattern builder once it's enabled.

(As an aside, I really miss XPath!)

fjeremic replied:
That sounds reasonable.

andrewcraik replied:
I think the disable of the matching is conflating the disable of OverflowCHK usage with the pattern matcher - we'll need to separate the two concepts to allow this stuff to be used elsewhere.

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -4986,7 +5014,15 @@ bool TR_EscapeAnalysis::fixupNode(TR::Node *node, TR::Node *parent, TR::NodeChec
                andNode->removeAllChildren();
 
-               if (TR::Compiler->target.is64Bit())
+               TR_ASSERT(andNode->getOpCodeValue() == TR::iand
```
This seems like it should be an assert fatal in the if ladder below processing andNode. Is there anything sensible to be done if we don't have an `iand` or `land` or do we need to bring the VM down?

hzongaro replied:
There's probably nothing sensible that we can do.  I'll rearrange the position of the assertion, and make it a `TR_ASSERT_FATAL`

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -3742,32 +3742,60 @@ static bool isFinalizableInlineTest(TR::Compilation *comp, TR::Node *candidate,
-          root->getFirstChild()->getFirstChild()->getFirstChild() == vftLoad /*&& (implied by the above assume)
+          r1->getOpCodeValue() == andOp &&
+          r2->getOpCodeValue() == constOp &&
+          (usesLongOps ? r2->getLongInt() == 0 : r2->getInt() == 0) &&
```
would this be more clear as:

`(usesLongOps ? r2->getLongInt() : r2->getInt()) == 0` - that would imply some kind of type conversion but that wouldn't affect the value 0 which is what we would be looking for...

andrewcraik commented on runtime/compiler/optimizer/EscapeAnalysis.cpp:
```diff
@@ -3742,32 +3742,60 @@ static bool isFinalizableInlineTest(TR::Compilation *comp, TR::Node *candidate,
+          (usesLongOps ? r2->getLongInt() == 0 : r2->getInt() == 0) &&
+          loadNode->getOpCodeValue() == loadOp &&
+          r12->getOpCodeValue() == constOp &&
+          (usesLongOps
```
could the == be factored out safely here as well or would the widening cause a problem?

andrewcraik commented:
@hzongaro any update since the issue this is fixing is tagged as 0.13?

hzongaro commented:
> @hzongaro any update since the issue this is fixing is tagged as 0.13?
In the process of running some sanity testing after addressing those comments

andrewcraik commented:
Please squash - I think that should address my concerns so I'll start sanity.

hzongaro commented:
Thanks, Andrew @andrewcraik .  Just squashed as commit https://github.com/eclipse/openj9/commit/dba4ac6da60bedfdaf29d102dc6cf64351e5f29f

andrewcraik commented:
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

andrewcraik commented:
sanity was an infra issue, the rest of the checks pass so merging - @hzongaro can you please create PR for the 0.13 release branch? FYI @pshipton
